### DISCUSSION TITLE
Add GetRelayLocationsByFilter mullvad-daemon

### DISCUSSION THREAD

**Discussion by denizt** [mullvad/mullvadvpn-app#9131] [Feature requests & Ideas]
Currently, [mullvad-daemon](https://github.com/mullvad/mullvadvpn-app/tree/master/mullvad-daemon) will return a big chunk of data (https://github.com/mullvad/mullvadvpn-app/blob/master/mullvad-management-interface/proto/management_interface.proto#L561) when getting the relay list via https://github.com/mullvad/mullvadvpn-app/blob/master/mullvad-management-interface/proto/management_interface.proto#L32 even if the consumer is only interested in certain sections.

Implement a new method:
`rpc GetRelayLocationsByFilter(RelayLocationsFilter) returns (RelayList) {}`
where the user can define the parameters they require, such as country, Wireguard, OpenVPN, etc..

Similar requests exist: https://github.com/mullvad/mullvadvpn-app/issues/2268, https://github.com/mullvad/mullvadvpn-app/issues/3136, https://github.com/mullvad/mullvadvpn-app/issues/3136 however have been open for some time. I wanted to check in before going ahead and adding this feature.

(Un)related? The daemon will just send the whole data, that is including OpenVPN locations, even if Wireguard only is chosen in settings, is this by design or should it be tracked as a separate issue?

**Comment by faern** [comment]
Is there any particular reason that your program receiving the list can't do the filtering? That's the way our current frontends are designed. They get all the settings from the daemon and they get all the servers. Then *if they want to* they filter out and show only a subset of the servers. It's a presentation issue and not a daemon issue IMO.

> Similar requests exist: https://github.com/mullvad/mullvadvpn-app/issues/2268, https://github.com/mullvad/mullvadvpn-app/issues/3136, https://github.com/mullvad/mullvadvpn-app/issues/3136 however have been open for some time. I wanted to check in before going ahead and adding this feature.

All of these are intended to be solved in the frontend. The last two ones are actually already done on desktop (you link to the same issue twice), because we added ownership filters in #3566.

> (Un)related? The daemon will just send the whole data, that is including OpenVPN locations, even if Wireguard only is chosen in settings, is this by design or should it be tracked as a separate issue?

This is indeed by design. Filtering is a presentation issue. If implemented like you suggests then no frontend program can know what OpenVPN relays are available without changing the setting. That would be limiting. Changing the setting causes a reconnect, maybe the frontend just wanted to know the available relays, not change any setting.

**Comment by denizt** [comment]
> Is there any particular reason that your program receiving the list can't do the filtering?

I am already able to do the filtering, however, when I initially pulled the relays, my client was not able to connect as 'Wireguard Only' was chosen in my preferences. So the official client and mine showed different relays. It also did not connect to any OpenVPN relay, suggesting no suitable relays were found. I found this error rather confusing. This also means the daemon knows the state.

The question here is if (all) the consumers want to do filtering, one way or another, wouldn't it be convenient to add it to the daemon? It would unify logic in one place, simplify implementation across clients (you already have 2, the CLI and the GUI), and provide, albeit negligible,  performance boost to the client. As you said, this is not something that can't be done on the client side, for the reasons I stated above, I think it would be beneficial to add it to the daemon. I also don't understand why adding a new method and not removing the old one would be limiting. And why would the filter be restricted to be in-line with the settings?

**Comment by faern** [comment]
> I also don't understand why adding a new method and not removing the old one would be limiting

We'd have to maintain functionality that we are not using ourselves. This means extra maintenance burden and we are quite likely to break it since we are not testing for it.

Other than that, I don't have much good arguments. We did it this way and it works. It's usually quite good to do filtering and sorting in the presentation layer. Yes, it requires more code, but it's also more flexible / allows the frontends to do more stuff.

**Comment by denizt** [comment]
The main rationale behind this request was to unify the code in daemon to simplify it for the clients. As you have mentioned, your client(s) already do this, and I think it would be good to move this code up the chain. As you build the API and are the sole user, I believe I have a different take on it, from a third-party developer perspective. For example, I don't see why all requests take Empty and there are different objects sent by the RelayList and the connect takes a different relay object. In my experience, at least, it's hard to make an API that pleases all the users 😄

I do appreciate that it would be extra work for you if you are not going to use it. I can implement filtering on my end.
